File: Binary Search/Guess Number Higher or Lower.cpp
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
public:
    int guessNumber(int n) {
        int l = 1;

        while (l <= n) {
            int mid = l + (n - l) / 2;
            switch (guess(mid)) {
                case 0:
                    return mid;
                    break;
                case 1:
                    l = mid + 1;
                    break;
                case -1:
                    n = mid - 1;
                    break;
            }
        }
        return 0;
    }
};

File: Binary Search/Guess Number Higher or Lower.rs
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * unsafe fn guess(num: i32) -> i32 {}
 */

impl Solution {
    unsafe fn guessNumber(n: i32) -> i32 {
        let mut l = 1;
        let mut r = n;

        while l <= r {
            let mid = l + (r - l) / 2;
            let res = guess(mid);
            if res == 0 {
                return mid;
            } else if res == 1 {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        -1
    }
}